src/simsopt/geo/permanent_magnet_grid.py

```diff
@@ -0,0 +1,541 @@
+                              like cartesian grid-alignment on a simple toroidal grid, which might
+                              be self-defeating.
+            pol_vectors:      Optional set of local coordinate systems for each dipole, 
+                              shape (Ndipoles, Ncoords, 3)
+                              which specifies which directions should be considered grid-aligned.
+                              Ncoords can be > 3, as in the PM4Stell design.
+    """
+
+    def __init__(
+        self, plasma_boundary,
+        rz_inner_surface=None,
+        rz_outer_surface=None, plasma_offset=0.1,
+        coil_offset=0.2, Bn=None, dr=0.1, dz=None,
+        Nx=10, Ny=None, Nz=None,
+        filename=None, surface_flag='vmec',
+        coordinate_flag='cartesian',
+        famus_filename=None,
+        pol_vectors=None
+    ):
+        if plasma_offset <= 0 or coil_offset <= 0:
+            raise ValueError('permanent magnets must be offset from the plasma')
+
+        if famus_filename is not None:
+            warnings.warn(
+                'famus_filename variable is set, so a pre-defined grid will be used. '
+                ' so that the following parameters are ignored: '
+                'rz_inner_surface, rz_outer_surface, dr, plasma_offset, '
+                'and coil_offset.'
+            )
+        self.famus_filename = famus_filename
+        self.filename = filename
+        self.surface_flag = surface_flag
+        self.plasma_offset = plasma_offset
+        self.coil_offset = coil_offset
+        self.Bn = Bn
+        self.dr = dr
+        self.Nx = Nx
+        if dz is not None:
+            self.dz = dz
+        else:
+            self.dz = dr
+        if Ny is not None:
+            self.Ny = Ny
+        else:
+            self.Ny = Nx
+        if Nz is not None:
+            self.Nz = Nz
+        else:
+            self.Nz = Nx
+
+        if coordinate_flag not in ['cartesian', 'cylindrical', 'toroidal']:
+            raise NotImplementedError(
+                'Only cartesian, cylindrical, and toroidal (simple toroidal)'
+                ' coordinate systems have been implemented so far.'
+            )
+        self.coordinate_flag = coordinate_flag
+
+        if not isinstance(plasma_boundary, SurfaceRZFourier):
+            raise ValueError(
+                'Plasma boundary must be specified as SurfaceRZFourier class.'
+            )
+        else:
+            self.plasma_boundary = plasma_boundary
+            self.R0 = plasma_boundary.get_rc(0, 0)
+            self.phi = self.plasma_boundary.quadpoints_phi
+            self.nphi = len(self.phi)
+            self.theta = self.plasma_boundary.quadpoints_theta
+            self.ntheta = len(self.theta)
+
+        if dr <= 0:
+            raise ValueError('dr grid spacing must be > 0')
+
+        t1 = time.time()
+        # If the inner surface is not specified, make default surface.
+        if (rz_inner_surface is None) and (famus_filename is None):
```

Comment:
This style (`if (rz_inner_surface is None) and (famus_filename is None)`) used to check for valid arguments is not pythonic. Remove the `famus_filename` argument and create a class method class `from_famus_file` and populate the arguments using famus_file in that classmethod.